cbits/is-valid-utf8.c

```diff
@@ -142,4 +153,544 @@ int bytestring_is_valid_utf8(uint8_t const *const src, size_t const len) {
+    // Load 128 bytes.
+    __m256i const *big_ptr = (__m256i const *)ptr;
+    __m256i const inputs[4] = {
+        _mm256_loadu_si256(big_ptr), _mm256_loadu_si256(big_ptr + 1),
+        _mm256_loadu_si256(big_ptr + 2), _mm256_loadu_si256(big_ptr + 3)};
+    // Check if we have ASCII.
+    bool is_ascii = _mm256_movemask_epi8(_mm256_or_si256(
+                        _mm256_or_si256(inputs[0], inputs[1]),
+                        _mm256_or_si256(inputs[2], inputs[3]))) == 0;
+    if (is_ascii) {
+      // Prev_first_len cheaply
+      prev_first_len =
+          _mm256_shuffle_epi8(first_len_tbl, high_nibbles_of_avx2(inputs[3]));
+    } else {
+      __m256i first_len =
+          _mm256_shuffle_epi8(first_len_tbl, high_nibbles_of_avx2(inputs[0]));
+      check_block_avx2(prev_input, prev_first_len, errors, first_range_tbl,
+                       range_min_tbl, range_max_tbl, df_ee_tbl, ef_fe_tbl,
+                       inputs[0], first_len);
+      prev_first_len = first_len;
+      first_len =
+          _mm256_shuffle_epi8(first_len_tbl, high_nibbles_of_avx2(inputs[1]));
+      check_block_avx2(inputs[0], prev_first_len, errors, first_range_tbl,
+                       range_min_tbl, range_max_tbl, df_ee_tbl, ef_fe_tbl,
+                       inputs[1], first_len);
+      prev_first_len = first_len;
+      first_len =
+          _mm256_shuffle_epi8(first_len_tbl, high_nibbles_of_avx2(inputs[2]));
+      check_block_avx2(inputs[1], prev_first_len, errors, first_range_tbl,
+                       range_min_tbl, range_max_tbl, df_ee_tbl, ef_fe_tbl,
+                       inputs[2], first_len);
+      prev_first_len = first_len;
+      first_len =
+          _mm256_shuffle_epi8(first_len_tbl, high_nibbles_of_avx2(inputs[3]));
+      check_block_avx2(inputs[2], prev_first_len, errors, first_range_tbl,
+                       range_min_tbl, range_max_tbl, df_ee_tbl, ef_fe_tbl,
+                       inputs[3], first_len);
+      prev_first_len = first_len;
+    }
+    // Set prev_input based on last block.
+    prev_input = inputs[3];
+    // Advance.
+    ptr += 128;
+  }
+  // Write out the error, check if it's OK.
+  __m256i const combined_errors = _mm256_or_si256(errors[0], errors[1]);
+  if (_mm256_testz_si256(combined_errors, combined_errors) != 1) {
+    return 0;
+  }
+  // 'Roll back' our pointer a little to prepare for a slow search of the rest.
+  uint32_t tokens_blob = _mm256_extract_epi32(prev_input, 7);
+  int8_t const *tokens = (int8_t const *)&tokens_blob;
+  ptrdiff_t lookahead = 0;
+  if (tokens[3] > (int8_t)0xBF) {
+    lookahead = 1;
+  } else if (tokens[2] > (int8_t)0xBF) {
+    lookahead = 2;
+  } else if (tokens[1] > (int8_t)0xBF) {
+    lookahead = 3;
+  }
+  uint8_t const *const small_ptr = ptr - lookahead;
+  size_t const small_len = remaining + lookahead;
+  return is_valid_utf8_fallback(small_ptr, small_len);
+}
+
+#endif
+
+#ifdef __x86_64__
+bool has_sse2() {
```

Comment:
These should really be marked `static inline`: we don't want these to link outside the module.

Reply:
Thanks!